barretenberg/cpp/src/barretenberg/vm/avm/trace/addressing_mode.hpp

```diff
@@ -47,26 +53,45 @@ template <size_t N> class Addressing {
         return Addressing<N>(modes, space_id);
     }
 
-    std::array<uint32_t, N> resolve(const std::array<uint32_t, N>& offsets, AvmMemTraceBuilder& mem_builder) const
+    AddressResolution<N> resolve(const std::array<uint32_t, N>& offsets, AvmMemTraceBuilder& mem_builder) const
     {
-        std::array<uint32_t, N> resolved;
+        std::array<uint32_t, N> resolved_addresses;
+
         for (size_t i = 0; i < N; i++) {
-            resolved[i] = offsets[i];
+            auto& res_addr = resolved_addresses[i];
+            res_addr = offsets[i];
             const auto mode = mode_per_operand[i];
             if ((static_cast<uint8_t>(mode) & static_cast<uint8_t>(AddressingMode::RELATIVE)) != 0) {
                 const auto mem_tag = mem_builder.unconstrained_get_memory_tag(space_id, 0);
-                // TODO(#9131): Error handling needs to be done
-                ASSERT(mem_tag == AvmMemoryTag::U32);
-                resolved[i] += static_cast<uint32_t>(mem_builder.unconstrained_read(space_id, 0));
+
+                if (mem_tag != AvmMemoryTag::U32) {
+                    return AddressResolution<N>{ .addresses = resolved_addresses,
+                                                 .error = AvmError::ADDR_RES_TAG_ERROR };
+                }
+
+                const auto base_addr = static_cast<uint32_t>(mem_builder.unconstrained_read(space_id, 0));
+
+                // Test if we overflow over uint32_t
+                if (res_addr + base_addr < base_addr) {
+                    return AddressResolution<N>{ .addresses = resolved_addresses,
+                                                 .error = AvmError::REL_ADDR_OUT_OF_RANGE };
+                }
+
+                res_addr += base_addr;
             }
+
             if ((static_cast<uint8_t>(mode) & static_cast<uint8_t>(AddressingMode::INDIRECT)) != 0) {
-                const auto mem_tag = mem_builder.unconstrained_get_memory_tag(space_id, resolved[i]);
-                // TODO(#9131): Error handling needs to be done
-                ASSERT(mem_tag == AvmMemoryTag::U32);
-                resolved[i] = static_cast<uint32_t>(mem_builder.unconstrained_read(space_id, resolved[i]));
+                const auto mem_tag = mem_builder.unconstrained_get_memory_tag(space_id, res_addr);
```

Comment:
Just like I mention in the TS file, it might be better to just throw OutOfRange in the mem_builder and catch in the opcode implementation